R/pmcmc.R

```diff
@@ -0,0 +1,428 @@
+  pars_min <- unlist(pars_min)  
+  pars_max <- unlist(pars_max)
+  pars_sd <- unlist(pars_sd)
+  pars_discrete <- unlist(pars_discrete)
+  
+  
+  
+  # create prior and likelihood functions given the inputs
+ 
+   if(is.null(log_prior)) {
+     # set improper, uninformative prior
+    log_prior <- function(pars) log(1e-10)
+   }
+  calc_lprior <- log_prior
+  
+  if(is.null(log_likelihood)) {
+    log_likelihood <- calc_loglikelihood
+  } else if (!('...' %in% names(formals(log_likelihood)))){
+    stop('log_likelihood function must be able to take unnamed arguments')
+  }
+  
+  
+  
+  # create shorthand function to calc ll given main inputs
+  calc_ll <- function(pars) {  
+    X <- log_likelihood(pars = pars, 
+                            data = data, 
+                            model_params = model_params,
+                            steps_per_day = steps_per_day, 
+                            pars_obs = pars_obs, 
+                            n_particles = n_particles
+    ) 
+    X
+  }
+  
+  
+  
+  # convert the current parameters into format easier for mcmc to deal with
+  curr_pars <- unlist(pars_init)
+  day_zero <- as.Date(data$date[1])
+  curr_pars[['start_date']] <- as.numeric(day_zero - pars_init$start_date) # convert to numeric
+  
+  
+  if(any(curr_pars < pars_min | curr_pars > pars_max)) {
+    stop('initial parameters are outside of specified range')
+  }
+  
+
+  ## calculate initial prior
+  curr_lprior <- calc_lprior(pars = curr_pars)
+  
+  # run particle filter on initial parameters
+  p_filter_est <- calc_ll(pars = curr_pars)
+  
+  
+  # checks on log_prior and log_likelihood functions
+  
+  if(length(curr_lprior) > 1) {
+    stop('log_prior must return a single numeric representing the log prior')
+  }
+  if(curr_lprior > 0 ) {
+    stop('log_prior must be negative or zero')
+  }
+  if(is.infinite(curr_lprior)) {
+    stop('initial parameters are not compatible with supplied prior')
+  }
+  
+  if(length(p_filter_est) != 2) {
+    stop('log_likelihood function must return a list containing elements log_likelihood and sample_state')
+  }
+  if(!setequal(names(p_filter_est), c('log_likelihood', 'sample_state'))) {
+    stop('log_likelihood function must return a list containing elements log_likelihood and sample_state')
+  }
+  if(length(p_filter_est$log_likelihood) > 1) {
+    stop('log_likelihood must be a single numeric representing the estimated log likelihood')
+  }
+  if(p_filter_est$log_likelihood > 0) {
+    stop('log_likelihood must be negative or zero')
+  }
+  if(any(p_filter_est$sample_state < 0)) {
+    stop('sample_state must be a vector of non-negative numbers')
+  }
+  
+  
+  # extract loglikelihhod estimate and sample state
+  # calculate posterior
+  curr_ll <- p_filter_est$log_likelihood
+  curr_lpost <- curr_lprior + curr_ll
+  
+  curr_ss <- p_filter_est$sample_state
+  
+
+  ## combine into a vector to add to the results at the end
```

Comment:
Can these just be added into `res` and `states` immediately? If you set `nrow = n_mcmc + 1`

Reply:
done and pushed